File: Search2DArray.cpp
// Time Complaxity  = O(log(n)) approximatley, since using this solution we syill have to check each row's end value.
// Space Complaxity =  O(1) I am not using any additinal space here.
// Runtime: 3 ms, faster than 84.36% of C++ online submissions for Search a 2D Matrix.
// Memory Usage: 9.6 MB, less than 5.13% of C++ online submissions for Search a 2D Matrix.



class Solution {
public:
    bool binarySearch (vector <int> arr, int target){
        int low = 0, high = arr.size()-1;
        
        while (low <= high) {
            int mid = (low+high)/2;
            
            if (arr[mid] == target) return true;
            
            if (arr[mid] > target)
                high = mid -1;
            else
                low = mid + 1;
        }
        return false;
    }
    
    bool searchMatrix(vector<vector<int>>& matrix, int target) {
        int m = matrix.size();
        if(m == 0) return false;
        int n = matrix[0].size();
        if(n == 0) return false;
        
        int row = 0, col = n-1;
        while(col >= 0 && row < m){
            if(matrix[row][col] == target){
                return true;
            }else if(target < matrix[row][col]){
                //the current column's value is larger than target so target must be in this row only if it is present
                if (binarySearch(matrix[row], target))
                    return true;
                else 
                    return false;
            }else{
                //the current row's value are all smaller than target
                row++;
            }
        }
        
        return false;
        
    }
};

File: search2DArray.go
// Time Complexity : O(log(m*n))
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : None

/*
 * Approach:
 * Since the guarantee for this problem is that any element in a row is smaller
 * than any element in the next row, we can consider this 2D array as a flat
 * sorted 1D array, and apply binary search on it. Note that the main change
 * here is the conversion of the "mid" into coordinates of the 2D array.
 */

package main

func searchMatrix(matrix [][]int, target int) bool {
	if len(matrix) == 0 || len(matrix[0]) == 0 {
		return false
	}

	m, n := len(matrix), len(matrix[0])

	low := 0
	high := m*n - 1

	for low <= high {
		mid := low + (high-low)/2 // prevent int overflow
		x, y := mid/n, mid%n
		if matrix[x][y] == target {
			return true
		}

		if matrix[x][y] < target { // move right
			low = mid + 1
		} else { // move left
			high = mid - 1
		}
	}

	return false
}
